ethcore/client-traits/src/lib.rs

```diff
@@ -413,7 +413,7 @@ pub trait StateClient {
 	type State: StateInfo;
 
 	/// Get a copy of the best block's state.
-	fn latest_state(&self) -> Self::State;
+	fn latest_state_and_header(&self) -> (Self::State, Header);
```

Comment:
I'm not a big fan of methods with `several responsibilities`. Why not a separate method `fn latest_header()`?

Is this some kind of optimization that I don't understand? 

Also, seems that we call this method a couple of this without caring about the header. I think the header type is quite big and might carry some cost to return by value when it's not needed.

Reply:
Looking up [latest state](https://github.com/paritytech/parity-ethereum/blob/4fd1ec643f1c64eef1770ac45deac6013b298bfd/ethcore/src/client/client.rs#L1036-L1045) first looks up latest header, so when we need both we:

- look up the header
- use the header to fetch the state
- toss away the header
- look up the header again

So while I agree with you that it's generally best to have methods do a single thing, I think returning both pieces of data is warranted here.

Reply:
I think it was made to ensure atomicity of the operation - i.e. that the we don't import the block in the meantime and the state does not match header.